File: external/grpc/scala/src/main/scala/org/aiddl/external/grpc/scala/sender/SenderClient.scala
package org.aiddl.external.grpc.scala.sender

import io.grpc.ManagedChannelBuilder
import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.parser.Parser
import org.aiddl.core.scala.representation.Term
import org.aiddl.external.grpc.aiddl.AiddlStr
import org.aiddl.external.grpc.scala.converter.Converter
import org.aiddl.external.grpc.sender.SenderGrpc

class SenderClient(host: String, port: Int, container: Container) {
  val converter = new Converter(container)
  val channel = ManagedChannelBuilder
    .forAddress(host, port)
    .usePlaintext()
    .build

  val blockingStub = SenderGrpc.blockingStub(channel)

  def send(x: Term): Unit = {
    blockingStub.send(this.converter.aiddl2pb(x))
  }

}

File: external/grpc/python/aiddl_external_grpc_python/sender/sender_client.py
import grpc

import aiddl_external_grpc_python.generated.sender_pb2_grpc as sender_pb2_grpc
from aiddl_external_grpc_python.converter import Converter


class SenderClient:
    def __init__(self, host, port, container):
        channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = sender_pb2_grpc.SenderStub(channel)
        self.converter = Converter(container)

    def Send(self, message):
        self.stub.send(self.converter.aiddl2pb(message))
